snuba_sdk/dsl/dsl.py

```diff
@@ -0,0 +1,309 @@
+    "+": ArithmeticFunction.PLUS.value,
+    "-": ArithmeticFunction.MINUS.value,
+}
+
+TERM_OPERATORS: Mapping[str, str] = {
+    "*": ArithmeticFunction.MULTIPLY.value,
+    "/": ArithmeticFunction.DIVIDE.value,
+}
+
+
+def parse_mql(mql: str) -> MetricsQuery:
+    """
+    Parse a MQL string into a MetricsQuery object.
+    """
+    try:
+        tree = GRAMMAR.parse(mql.strip())
+    except ParseError as e:
+        raise InvalidQueryError("Invalid metrics syntax") from e
+    result = MQLlVisitor().visit(tree)
+    assert isinstance(result, (Timeseries, Formula))
+    metrics_query = MetricsQuery(query=result)
+    return metrics_query
+
+
+class MQLlVisitor(NodeVisitor):
+    def visit(self, node: Node) -> Any:
+        """Walk a parse tree, transforming it into a MetricsQuery object.
+
+        Recursively descend a parse tree, dispatching to the method named after
+        the rule in the :class:`~parsimonious.grammar.Grammar` that produced
+        each node. If, for example, a rule was... ::
+
+            bold = '<b>'
+
+        ...the ``visit_bold()`` method would be called.
+        """
+        method = getattr(self, "visit_" + node.expr_name, self.generic_visit)
+        try:
+            result = method(node, [self.visit(n) for n in node])
+            return result
+        except Exception as e:
+            raise e
+
+    def collapse_into_timeseries(self, formula: Formula) -> Timeseries:
+        """
+        Collapses the filters and groupbys of a MetricsQuery into the Timeseries object
+        using the distributive property.
+
+        For example:
+        (sum(foo){tag="value"} / sum(bar){tag="value"}){tag2="value2"}
+        becomes ->
+        (sum(foo){tag="value", tag2="value2"} / sum(bar){tag="value", tag2="value2"})
+        """
+        formula_filters = formula.filters if formula.filters else []
+        formula_groupby = formula.groupby if formula.groupby else []
+        new_parameters = []
+        if formula.parameters:
+            for parameter in formula.parameters:
+                if isinstance(parameter, Timeseries):
+                    timeseries_filters = parameter.filters if parameter.filters else []
+                    timeseries_groupby = parameter.groupby if parameter.groupby else []
+                    combined_filters = formula_filters + timeseries_filters
+                    combined_groupby = formula_groupby + timeseries_groupby
+                    combined_filters = combined_filters if combined_filters else None
+                    combined_groupby = combined_groupby if combined_groupby else None
+                    new_timeseries = parameter.set_filters(combined_filters).set_groupby(combined_groupby)
+                    new_parameters.append(new_timeseries)
+                else:
+                    new_parameters.append(parameter)
+            formula.parameters = new_parameters
+
+
+
+    def visit_expression(self, node: Node, children: Sequence[Any]) -> Any:
```

Comment:
All of these functions should be typed specifically. See the [SNQL parser](https://github.com/getsentry/snuba/blob/master/snuba/query/snql/parser.py#L325) for how we do that. We also type the children sequence as much as possible.